plan/predicate_push_down.go

```diff
@@ -47,10 +48,248 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
+	//    ATTENTION: here column 'e' doesn't belong to any mep, so we skip "e != 0".
+	// 3. propagate constants in these inequality predicates, and we finally get:
+	//    "a = b and c = d and a = c and e = f and g = h and e != 0 and a > 0 and b > 0 and c > 0 and d > 0 and g like 'abc' and h like 'abc' ".
+	multipleEqualities := make(map[*expression.Column]*expression.Column, 0)
+	union := func(leftExpr *expression.Column, rightExpr *expression.Column) {
+		rootOfLeftExpr, ok1 := multipleEqualities[leftExpr]
+		rootOfRightExpr, ok2 := multipleEqualities[rightExpr]
+		if !ok1 && !ok2 {
+			multipleEqualities[leftExpr] = leftExpr
+			multipleEqualities[rightExpr] = leftExpr
+		} else if ok1 && !ok2 {
+			multipleEqualities[rightExpr] = rootOfLeftExpr
+		} else if !ok1 && ok2 {
+			multipleEqualities[leftExpr] = rootOfRightExpr
+		} else if !rootOfLeftExpr.Equal(rootOfRightExpr) {
+			for k, v := range multipleEqualities {
+				if v.Equal(rootOfRightExpr) {
+					multipleEqualities[k] = v
+				}
+			}
+		}
+	}
+	for _, cond := range conditions { // build multiple equality predicates.
+		expr, ok := cond.(*expression.ScalarFunction)
+		if !ok || expr.FuncName.L != ast.EQ {
+			continue
+		}
+		left, ok1 := expr.Args[0].(*expression.Column)
+		right, ok2 := expr.Args[1].(*expression.Column)
+		if ok1 && ok2 {
+			union(left, right)
+		}
+	}
+	if len(multipleEqualities) == 0 {
+		return conditions
+	}
+	inequalityFuncs := map[string]string{
+		ast.LT:   ast.LT,
+		ast.GT:   ast.GT,
+		ast.LE:   ast.LE,
+		ast.GE:   ast.GE,
+		ast.NE:   ast.NE,
+		ast.Like: ast.Like,
+	}
+	type transitiveInEqualityPredicate map[string]map[int]map[string][]*expression.Constant // transitive inequality predicates between one column and one constant.
+	inequalities := make(transitiveInEqualityPredicate, 0)
+	for i := 0; i < len(conditions); i++ { // extract inequality predicates.
+		expr, ok := conditions[i].(*expression.ScalarFunction)
+		if !ok {
+			continue
+		}
+		var (
+			col      *expression.Column
+			equalCol *expression.Column // the root column corresponding to a column in a multiple equality predicate.
+			fromID   string
+			position int
+			val      *expression.Constant
+			isColumn bool
+			funcName string
+		)
+		left, ok1 := expr.Args[0].(*expression.Constant)
+		right, ok2 := expr.Args[1].(*expression.Constant)
+		if col, isColumn = expr.Args[0].(*expression.Column); ok2 && isColumn {
+			val = right
+		} else if col, isColumn = expr.Args[1].(*expression.Column); ok1 && isColumn {
+			val = left
+		} else {
+			continue
+		}
+		funcName, ok = inequalityFuncs[expr.FuncName.L]
+		if !ok {
+			continue
+		}
+		equalCol, ok = multipleEqualities[col]
+		if !ok { // no need to propagate inequality predicates whose column is only equal to itself.
+			continue
+		}
+		fromID, position = equalCol.FromID, equalCol.Position
+		if _, ok = inequalities[fromID]; !ok {
+			inequalities[fromID] = make(map[int]map[string][]*expression.Constant, 1)
```

Comment:
It's better to encode (FromID, PositionID) as a HashCode binary key by using EncodeValue